MR-2755: Add data flow diagrams (#1311)
CMSgov/managed-care-review: docs/technical-design/form-data-flow.md

# Form-data flow

The MC-Review site's primary function is to collect form data from states, and package it for review by CMS. In this document, we'll describe the flow of that form data through the app; from creation to storage, and back from storage to the UI.

## Information entry

```mermaid
graph TD
    A["User enters form data"] -->|"validated by Formik and Yup"| B["Data submit (by clicking 'continue' or 'save as draft')"]
    B -->|"convert data to a base64 protocol buffer<br />(toProtoBuffer)"| C["GraphQL resolver"]

>     C -->|"convert data back to domain models for various checks<br />(toDomain)"| D["Prisma (postgres)"]

Reviewer: ```suggestion
    C -->|"convert data back to  backend domain models for various checks<br />(toDomain)"| D["Prisma (postgres)"]
```

Added the work "backend" to domain models since we have both frontend and backend domain models

    D -->|"convert again to protobuf"| E["Written to database"]
```

## Information retrieval

```mermaid
graph TD
    A["User clicks link to a package"] -->B["GraphQL resolver"]
    B --> C["Query Prisma (postgres)"]
    C -->|"data is returned to the resolver for checking" | D["(the same) GraphQL Resolver"]

>     D -->|"run various checks"| E["Front-end"]

Reviewer: ```suggestion
    D -->|"run various checks and convert to back to protocol buffer"| E["Front-end"]
```

Author: > Zod is not used on the frontend

@haworku Are we using different definitions of "frontend?"  Zod is used in `toDomain`.  [Here](https://github.com/CMSgov/managed-care-review/blob/main/services/app-web/src/common-code/proto/healthPlanFormDataProto/toDomain.ts#L443).

Author: We chatted about this offline; it's ok to keep it as-is.

>     E -->|"convert from protocol buffers to Typescript domain models<br />(toDomain)<br />validate with Zod"| F["UI"]

Reviewer: ```suggestion
    E -->|"convert from protocol buffers to frontend domain models"| F["UI"]
```

```

### Supplements

For an overview of what the data is like, and why it's being converted to and from protocol buffers, see [the serialization ADR](../architectural-decision-records/008-form-data-serialization.md).

> 

Reviewer: Let's add the schemas themselves as supplement files! we can link:

schemas
- proto schema
- zod schema
- formik schemas (or show how to find them)

Data transformation functions
- toDomain
- toProtobuffer
